Add chapter details

hkamran80 opened the pull request:
This PR adds chapter details, like the index, name, and summary, to the work summary object. It's exposed through the `getWorkChapter` function and the `ChapterWorkSummary` interface.

Tests are also included, this time with handlers and mock data!

enigmalea commented:
Thanks so much for your pull request!

@essential-randomness asked me to reply because she's a fool who trusts me to maintain repos and wants me to practice. (I really don't know what she's thinking.)

What do you think about adding the chapter info to the existing `getWork` function instead of creating a separate `getWorkChapter` function? The thought is to update `getWork(workId)` to accept parameters `{workId, chapterId}` where chapterId is optional:
```typescript
export const getWork = async ({
  workId,
 chapterId
}: {
  workId: string;
  chapterId?: string;
}): Promise<WorkSummary | LockedWorkSummary> => {
```

Then we would add a chapterInfo key to the WorkSummary interface which is null for works with one chapter and which contains chapter info for works with multiple chapters:
```typescript
export interface WorkSummary {
// the rest of the interface
  chapterInfo: {
    id: string;
    index: number;
    name: string | null;
    summary: string | null;
  } | null;
}
```

Since a work can have both a summary for the entire work and one for each chapter, it should be fine to return both, if they both exist. This would allow the end user to decide whether they want to use `Work.summary` or `Work.chapterInfo.summary` in their use case.

hkamran80 commented:
I think you're doing a wonderful job at maintaining! That's a great idea! I'll implement it right now!

hkamran80 commented:
The reason I removed `WorkSummary.summary` from `ChapterWorkSummary` is because the first chapter of a work cannot have a chapter summary and a work summary to my knowledge. Do you know of any fics that have both?

enigmalea commented:
The reason you don't see it often is that the first time you create a work the summary box automatically is the work summary, but if you go back to edit Chapter 1 after Chapter 2 is created, another blank summary is present for editing. I discovered this by _accidentally_ creating summaries under Chapter 1 a couple of times and having to go back to delete them.

I'm not sure I've ever seen a fic with both intentionally, but it _is_ possible.

hkamran80 commented:
The code just picks all the summaries with the `.summary blockquote.userstuff` selector, so both should be selected in that case.

enigmalea commented on src/works/index.ts:
```diff
@@ -69,6 +78,15 @@ export const getWork = async ({
     publishedAt: getWorkPublishDate(workPage),
     updatedAt: getWorkUpdateDate(workPage),
+    chapter:
+      chapterId && totalChapters !== 1
```
I've been thinking about this a bit, and I think that the `&& totalChapters !== 1` isn't necessary. 

AO3 has three chapter configurations:
- #/? - these might have a chapterId provided by AO3 if there's more than one chapter, but won't if there's only one chapter. 
- 1/1 - This is a single chapter fic aka one shot. These are not provided a chapterId by AO3.
- #/# - where #> 1 in both cases. All chapters will have a chapterID.

Since AO3 doesn't provide chapterIDs if there's only one chapter, I think we can simplify this to just `chapterId` what do you think?

hkamran80 replied:
That will work. The `totalChapters` check was the way I was originally checking if the object should be `null` or not. I'll update it.

hkamran80 replied:
Actually, I remembered why the `totalChapters` check was still there. It's so a chapter summary can still be returned for a multi-chapter fic, even if no `chapterId` is passed.

enigmalea commented on src/works/index.ts:
```diff
@@ -69,6 +78,15 @@ export const getWork = async ({
     },
     publishedAt: getWorkPublishDate(workPage),
     updatedAt: getWorkUpdateDate(workPage),
+    chapter:
```
I think we probably want to call this `chapterInfo` to be more descriptive for what the data it actually holds. This will be super useful for end users who are using typescript and are getting hints from their editor.

enigmalea reviewed (commented):
Thanks so much for the changes. 

In addition to the two small suggestions I made, I was thinking about our WorkSummary versus ChapterSummary conundrum.

I updated [one of my fics](https://archiveofourown.org/works/17793689/chapters/41980418) to have a chapter summary and work summary on chapter 1.

I haven't tested it (feel free to use my work as a mock!), but I think the current getWorkSummary function might return both summaries as one string, which is less than ideal behavior.

We might need to somehow update getWorkSummary to exclude the chapter summary text, and create a separate getChapterSummary function that selects specifically for `.chapter .summary blockquote.userstuff`.

essential-randomness commented:
Pinging @enigmalea!

Let me know if you're busy and i'll take care of this :)

essential-randomness commented on src/page-loaders.ts:
```diff
@@ -50,10 +50,10 @@ export const loadTagFeedAtomPage = async ({ tagId }: { tagId: string }) => {
   kind: "WorkPage";
 }
-export const loadWorkPage = async (workId: string) => {
+export const loadWorkPage = async (workId: string, chapterId?: string) => {
```
Use the object params style as we do elsewhere, since it makes codes more readable and is more flexible when it comes to adding additional params:

```ts
export const loadWorkPage = async (params: { workId: string, chapterId?: string}) => {
```

essential-randomness reviewed (changes requested):
I left one small comment on the function signature for `loadWorkPage`, and i think all we're missing is what @enigmalea (who's sick, so I'm taking over) was saying about chapter one summary being different from chapter summary (as seen in her work: https://archiveofourown.org/works/17793689/chapters/41980418).

I would also write a test that use that fic chapter as a mock, cause i don't know if you get any error with the current code when there's more than one result returned for the same query selector.

If you don't want to make this change now, feel free to add a todo mentioning the problem and linking to the chapter and we'll get to it later.

hkamran80 commented:
After @enigmalea mentioned that chapter one could have a work and chapter summary, I added a separate function (`getChapterSummary`) which only gets the chapter's summary. I also modified the `getWorkSummary` function to only get the work's summary by checking for the `preface` class.

I'll update the function signature for `loadWorkPage` and add tests in a bit.

essential-randomness commented:
Woops, sorry about that, clearly I looked at it too fast and didn't see the logic had been updated (it's been a long day)! Tests will help clarify :)

hkamran80 commented:
All good!

hkamran80 commented:
If this is ready for a merge, I recommend using a squash commit.

essential-randomness reviewed (approved):
Amazing!!! Thank you so much once again for the excellent contribution. Squash merging and releasing now.

hkamran80 commented:
Happy to help!!
